feature: move-form

snoe opened the pull request:
Mostly working:
Reworked add-to-namespace add a seam where a more explicit libspec could
be used.
Added a nice test helper to turn a set of changes into a document as a
client would.

Move form incomplete pieces:
- needs to add missing lib for all namespaces used within form
- needs to handle whitespace better
- removing the source require erroneously

- [x] I created a issue to discuss the problem I am trying to solve or there is already a open issue. #566 
- [x] I added a new entry to [CHANGELOG.md](https://github.com/clojure-lsp/clojure-lsp/blob/master/CHANGELOG.md)
- [x] I updated documentation if applicable (`docs` folder)

snoe commented:
```
FAIL in clojure-lsp.feature.add-missing-libspec-test/add-import-to-namespace-test (add_missing_libspec_test.clj:354)
when on an invalid location
expected: (nil? (-> (h/code "(ns foo.bar) |;; comment") (add-import-to-namespace "java.util.Date")))
  actual: (not (nil? ({:range {:row 1, :col 1, :end-row 1, :end-col 13}, :loc [<list:
  (ns foo.bar
    (:import
     java.util.Date))
> {:l [], :pnodes [<forms:
  (ns foo.bar
    (:import
      java.util.Date))
>], :ppath nil, :r nil, :changed? true}]})))
1 tests, 8 assertions, 1 failures.
```

@mainej I don't understand the purpose of this test, afaict the position of the cursor is inconsequential to the addition.

snoe commented:
@ericdallo I kind of want to add this sooner than later to avoid merge hell.
I believe the foundation is in place and it can be iterated on & more tests can be added in future PRs. So maybe don't advertise this initial pr or mark it as alpha or something.


here's how I'm calling it from vim: 
`nnoremap <silent> crmf :call CocRequest('clojure-lsp', 'workspace/executeCommand', {'command': 'move-form', 'arguments': [Expand('%:p'), line('.') - 1, col('.') - 1, input("File: ", "", "file")]})<CR>`

I want to spend another day on getting the checklist mostly done, especially the last point before taking this out of draft.

ericdallo commented:
Thanks @snoe! I'm releasing a new version right now and thought would be good to keep this PR and the logging one outside of the release to have more time to test if we find bugs soon, but we can merge this on master as soon as you finish/we review and test :)

mainej commented:
@snoe at one point I wrote a bunch of tests about invoking the action menu and/or commands from a top-level comment or whitespace. This caught a ton of problems, some of them new (from #716) and some of them old. As I was adding the tests, honestly, I wasn't concerned with _what_ the command did, just that it didn't do anything completely nonsensical or throw any exceptions. I'd prefer to keep some version of these tests to prevent regressions, but feel free to make them specify different behavior.

snoe commented:
@ericdallo This should be ready for review, I marked it as ALPHA in the changelog, but it seems pretty solid, I dogfooded quite a bit.  

### Future work considerations.
It's very strict and some feedback to the user would be very helpful.
- the destination file must exist with an ns form, 
- the cursor must be at the start paren of the form to move. 
- the form being moved cannot use any vars from the source namespace.

It does not check for cyclic dependencies, I believe that should be added somewhere around `add-to-namespace` no matter the command.

### Other stuff

My vim command for invoking is here, I chose a relative filename to the project-root as the arg because I assume most editors should be able to provide completion here.
`nnoremap <silent> crmf :call CocRequest('clojure-lsp', 'workspace/executeCommand', {'command': 'move-form', 'arguments': [Expand('%:p'), line('.') - 1, col('.') - 1, input("File: ", "", "file")]})<CR>`

I addressed much of #828 in features.md as I added the new command doc.

ericdallo commented:
Sounds great @snoe , I'll take a look today and test it!

> the destination file must exist with an ns form,
the cursor must be at the start paren of the form to move.
the form being moved cannot use any vars from the source namespace.

I think for now this is ok, in the future this should not be hard to make work, right? at least the first and the second

snoe commented:
> Sounds great @snoe , I'll take a look today and test it!
> 
> > the destination file must exist with an ns form,
> > the cursor must be at the start paren of the form to move.
> > the form being moved cannot use any vars from the source namespace.
> 
> I think for now this is ok, in the future this should not be hard to make work, right? at least the first and the second

Yup, should be pretty straightforwad

ericdallo commented on docs/features.md:
```diff
@@ -388,38 +374,51 @@ For more information check the [stubs settings section](../settings.md#stub-gene
+| ai       | add-import-to-namespace | Add import to namespace                | `[file-uri,row,col[,name]]`                      | ✅                         | <img src="../images/features/add-common-import.gif" width=360>       |
+| am       | add-missing-libspec     | Add missing require                    | `[file-uri,row,col]`                             | ✅                         | <img src="../images/features/add-missing-require.gif" width=360>     |
+| as       | add-require-suggestion  | Add require suggestion                 | `[file-uri,row,col,ns,alias,refer]`              | ✅                         | <img src="../images/features/add-alias-suggestion.gif" width=360>    |
+| cc       | cycle-coll              | Cycle collection `(#{}, {}, [], ())`   | `[file-uri,row,col]`                             |                           |                                                                      |
```
```suggestion
| cc       | cycle-coll              | Cycle collection `(#{}, {}, [], ())`   | `[file-uri,row,col]`                             |      ✅                   |                                                                      |
```

ericdallo commented on lib/src/clojure_lsp/parser.clj:
```diff
@@ -11,7 +11,7 @@
 #_{:clj-kondo/ignore [:clojure-lsp/unused-public-var]}
 (defmacro zspy [loc]
   `(do
-     (log/warn '~loc (pr-str (z/sexpr ~loc)))
+     (taoensso.timbre/warn '~loc (pr-str (z/sexpr ~loc)))
      ~loc))
```
I wonder if we really need to add timbre to test alias just for this function, I don't use this at all as it just replace a str or a expr

snoe replied:
The reason to add it to the test alias is so you can actually log through koacha, not because of this macro.

ericdallo commented on docs/features.md:
```diff
@@ -300,27 +306,7 @@ User can register additional custom snippets, for more information on how to con
-| Sort map                           | <img src="../images/features/sort-map.gif" width=720>                |
-| Suppress diagnostic                |                                                                      |
-| Create test for function           |                                                                      |
+See [below](#clojuser-lsp-extra-commands) for screenshots.
```
IMO we should mention both commands and code actions but give more highlights for code actions, as they are smarter than commands, as they are available when they make sense be available.

mainej replied:
I agree. Or rather, I'd say that @snoe, the wording of the new section strongly emphasizes binding commands, before offhandedly mentioning code actions. I think it'd be a barrier to entry if everyone thought they had to set up their own keybindings to use clojure-lsp. I know you use keybindings more, and that's sort of the vim way. And I like that way too, because I use Emacs in evil mode. But I'd guess that the majority of users are going to favor the code actions. And what @ericdallo has been saying is very true... the code actions guide everyone, but particularly new users, to refactorings that are appropriate for the current cursor position. So, I'd suggest explaining code actions before invoking the "we **strongly** suggest" language for setting up keybindings.

snoe replied:
Here's the thing, code actions don't really need documentation from us in the way that executeCommand does. It falls in the same bucket as go to def and rename; lsp client features.

Fair enough point about the ordering, so I moved the note about code actions above the keybindings paragraph.

ericdallo replied:
I see, even so, as a new clojure-lsp user, I would like to know what features clojure-lsp offer, and mentioning only commands may leave users think this would work only manually.
Well, I'm ok leaving the note as the first thing though

ericdallo commented on lib/src/clojure_lsp/feature/move_form.clj:
```diff
@@ -0,0 +1,203 @@
+                       multiple-defs?)]
+    (when can-move?
+      (let [def-to-move (first defs)
+            refs (q/find-references analysis def-to-move false true)
```
What is this `true`? shouldn't we pass db there?

snoe replied:
I think I had started down the path of only looking for project refs. I see @mainej has added something similar, and this can be updated later if desired. It's generally less important as it's ok imo, for this command to take slightly longer.

ericdallo replied:
@snoe I mean, that find-references function, receives as last arg a `db` right? Why we are passing true instead of a db component?

snoe replied:
right, it was a good catch, db is being passed now.

ericdallo commented on docs/features.md:
```diff
@@ -388,38 +374,51 @@ For more information check the [stubs settings section](../settings.md#stub-gene
+| is       | inline-symbol           | Inline Symbol                          | `[file-uri,row,col]`                             | ✅                         | <img src="../images/features/inline-symbol.gif" width=360>           |
+| ma       | resolve-macro-as        | Resolve macro as                       | `[file-uri,row,col]`                             | ✅                         | <img src="../images/features/resolve-macro-as.gif" width=720>        |
+| md       | move-coll-entry-down    | Move coll entry down                   | `[file-uri,row,col]`                             | ✅                         | <img src="../images/features/move-coll-entry.gif" width=720>         |
+| mf       | move-form               | Move form                              | `[file-uri,row,col,filename]`                    | ✅                         |                                                                      |
```
I was wondering, maybe we could receive `file-uri, row, col, dest-namespace`, that would be a better UX?
Also, I though about not receiving that at all, and asking it so a future code action could use the same command, similar to how we do with `resolve-macro-as` and `create-test`, WDYT?

snoe replied:
I used the dest-filename because editors can autocomplete it very easily.

ericdallo replied:
I think receiving a `ns` has a better semantic for the command as it doesn't make sense to move to a file outside the project, or not in the source-path?
We do similar on `ns` renames, we ask for the new ns name and not the filename.
I know editors can complete filenames, but as a IDE maybe we should restrict the input to make even clear what we expect to user input

snoe replied:
I think ns has a ton of ambiguity that files don't.
Macros in cljs often have two files with the same namespace, one .clj file one .cljs file.
If we want to extend this to create a new ns, you'd still have to ask for the path.
Rename can assume the same src folder, whereas this would often be used to move from .clj to a shared .cljc src dir.

If you wanted, you can add a restriction that checks that the filename is within one of the source paths.

ericdallo replied:
I see, makes sense I think

ericdallo commented:
@snoe could you solve conflicts one last time so I can finish the review and test it manually?

mainej commented on docs/features.md:
```diff
@@ -388,38 +374,51 @@ For more information check the [stubs settings section](../settings.md#stub-gene
-
-__Note: Most of them are available via code actions and most of the time you want to call the code action and not the command manually__
-
-- Clean namespace *
```
Glad to see these two nearly identical lists merged into one!

snoe commented:
@ericdallo I believe this is ready.

ericdallo commented:
@snoe , thanks, just a couple of more questions/suggestions, LMK what you think

ericdallo commented:
@snoe I tested and it looks good, some things to discuss for next work on that feature:

- any reason check if zloc is the top-form one? could we just get top-form always and use that?
- I noticed we don't move/copy the requires of the inner usages, something to improve later
- I thought a little about how to make that work with code actions and I didn't find a way to ask for the dest-file via clojure-lsp :/ I know it's possible to make emacs/lsp-mode intercept that, but all clients would need to do that, and some of them probably don't have that flexibility, for some code actions we hard-code a string like extract-function, but this one I can't see it working

snoe commented:
> any reason check if zloc is the top-form one? could we just get top-form always and use that?

Yes, probably.

> I noticed we don't move/copy the requires of the inner usages, something to improve later

agree.

> code actions

Yeah, it's more similar to a rename, maybe a code-action specific command like find-suggestions where the server asks client for the file and kicks off the "real" command

ericdallo commented:
> Yeah, it's more similar to a rename, maybe a code-action specific command like find-suggestions where the server asks client for the file and kicks off the "real" command

Good suggestion, I'll think in something about that later
